### PR TITLE
#165273589 Feature to allow users to report an article

### PR DESCRIPTION
**Pull request by llwasampijja**
#### What does this PR do?
Reports an article which violates the terms of use of the platform
#### Description of Task to be completed?
A user can report an article, update the report for an article, delete a report of an article, list reports for an article, fetch a report of an article.
#### How should this be manually tested?
* Use postman [this](https://documenter.getpostman.com/view/5689256/S1LsZqkx) postman collection to test these endpoints
#### Any background context you want to provide?
Readme file
#### What are the relevant pivotal tracker stories?
[#165273589](https://www.pivotaltracker.com/story/show/165273589)
#### Screenshots (if appropriate)
None
#### Questions:
None

### CODE REVIEW

File: authors/apps/report/models.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from authors import settings
+from authors.apps.articles.models import Article
+from authors.apps.profiles.models import Profile
+
+# Create your models here.
+class ReportArticle(models.Model):
+    """model for reporting an article"""
+    reporter = models.ForeignKey(Profile, on_delete=models.CASCADE)
+    article = models.ForeignKey(Article, to_field="slug", on_delete=models.CASCADE)
+    violation_subject = models.CharField(max_length=100, blank=False, null=False)
```
Comment (kalsmic) [inline authors/apps/report/models.py]: well done @llwasampijja, I was able to report a violation and also noticed that I was not able to report my own article which is good. However, I was thinking of the violation_subject field as a ChoiceField like how it is on medium.com where the user selects the category of the violation for which they are reporting the article. We predefine the categories for violations.

**Review by bisonlou — CHANGES_REQUESTED** [review]
I have noticed your app name is singular. However, according to [best practices](https://pt.slideshare.net/accavdar/django-best-practices-46944311) #39, an app name should be plural.

File: authors/apps/report/tests/test_report.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+from authors.apps.authentication.models import User
+from authors.apps.report.tests.base_test import BaseTest
+
+class TestArticleReport(BaseTest):
+
+    def test_post_report_success(self):
+        response = self.client.post(
+            self.report_article_url_post, data=self.report_data, format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (kalsmic) [inline authors/apps/report/tests/test_report.py]: Please test for the data at least for the success.

**Review by kalsmic — CHANGES_REQUESTED** [review]

File: authors/apps/report/views.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            reported_article = ReportArticle.objects.get(id=kwargs.get("report_id"))
+            serializer = self.get_serializer(reported_article)
+            current_user = Profile.objects.get(user=request.user)
+            if request.user.is_superuser or reported_article.reporter == current_user:
+                return Response ({"report":serializer.data})
+            else:
+                return Response (
+                    {"error":"You don't have the permissions to perform this request"},
+                    status=status.HTTP_401_UNAUTHORIZED
+                )
```
Comment (ZackAtama) [inline authors/apps/report/views.py]: Great work @llwasampijja, however, the `HTTP_401_UNAUTHORIZED` status code should be used explicitly for users that aren't logged in, in this case a user has access to the platform however, they aren't allowed to access this route, so the `HTTP_403_FORBIDDEN` status code best fits this permission check. Thank you.

**Review by ZackAtama — CHANGES_REQUESTED** [review]

File: authors/apps/report/views.py (hunk 3 above)
Comment (llwasampijja) [reply authors/apps/report/views.py]: @AtamaZack ok, let me change that!

**Review by ZackAtama — APPROVED** [review]

**Review by bisonlou — APPROVED** [review]

**Review by kalsmic — APPROVED** [review]

**Pull request merged**